pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -97,10 +101,85 @@ func (c *clusterMemberRemovalController) sync(ctx context.Context, _ factory.Syn
 	if err := c.attemptToRemoveLearningMember(ctx); err != nil {
 		errs = append(errs, err)
 	}
+	if err := c.attemptToScaleDown(ctx); err != nil {
+		errs = append(errs, err)
+	}
 	return kerrors.NewAggregate(errs)
 
 }
 
+// attemptToScaleDown attempts to remove a voting member only once we have identified that
+// a Machine resource is being deleted and a replacement member has been created
+func (c *clusterMemberRemovalController) attemptToScaleDown(ctx context.Context) error {
+	currentVotingMemberIPListSet, err := c.votingMemberIPListSet()
+	if err != nil {
+		return err
+	}
+
+	desiredControlPlaneReplicasCount, err := ceohelpers.ReadDesiredControlPlaneReplicasCount(c.operatorClient)
+	if err != nil {
+		return err
+	}
+	if desiredControlPlaneReplicasCount == 0 {
+		return fmt.Errorf("desired control plane replicas count cannot be empty")
+	}
+	if currentVotingMemberIPListSet.Len() <= desiredControlPlaneReplicasCount {
+		klog.V(4).Infof("Ignoring scale-down since the number of etcd voting members (%d) < desired number of control-plane replicas (%d) ", currentVotingMemberIPListSet.Len(), desiredControlPlaneReplicasCount)
+		return nil
+	}
+
+	memberMachines, err := c.currentMemberMachines()
+	if err != nil {
+		return err
+	}
+
+	var votingMachines []*machinev1beta1.Machine
+	for memberMachineIP, memberMachine := range ceohelpers.IndexMachinesByNodeInternalIP(memberMachines) {
+		if currentVotingMemberIPListSet.Has(memberMachineIP) {
+			votingMachines = append(votingMachines, memberMachine)
+		}
+	}
+
+	votingMachinesPendingDeletion := ceohelpers.FilterMachinesPendingDeletion(votingMachines)
+	if len(votingMachinesPendingDeletion) == 0 {
+		return nil
+	}
+
+	// based on the data in the cache it looks like we have something to do
+	// get the live members, to observe the current state and the member IDs
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return err
+	}
+
+	var liveVotingMembers []*etcdserverpb.Member
+	for _, member := range members {
+		if member.IsLearner {
+			continue
+		}
+		liveVotingMembers = append(liveVotingMembers, member)
+	}
+
+	// do not trust data in the cache, compare with the current state
+	if len(liveVotingMembers) <= desiredControlPlaneReplicasCount {
+		klog.V(2).Infof("Ignoring scale-down since the number of live etcd voting members (%d) < desired number of control-plane replicas (%d) ", len(liveVotingMembers), desiredControlPlaneReplicasCount)
```

Comment:
yeah, I think before this effort is done, you'll need to create way to "write this event if it has been more than five minutes since I last wrote it"